pkg/history/populate_headers.go

```diff
@@ -24,13 +24,15 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-func PopulateMissingHeaders(blockChain core.BlockChain, headerRepository datastore.HeaderRepository, startingBlockNumber int64) (int, error) {
-	lastBlock, err := blockChain.LastBlock()
+func PopulateMissingHeaders(blockChain core.BlockChain, headerRepository datastore.HeaderRepository, startingBlockNumber,
+	validationWindow int64) (int, error) {
+	chainHead, err := blockChain.ChainHead()
 	if err != nil {
 		return 0, fmt.Errorf("error getting last block: %w", err)
 	}
 
-	blockNumbers, err := headerRepository.MissingBlockNumbers(startingBlockNumber, lastBlock.Int64())
+	lastBlock := getLastBlock(startingBlockNumber, chainHead.Int64(), validationWindow)
+	blockNumbers, err := headerRepository.MissingBlockNumbers(startingBlockNumber, lastBlock)
```

Comment:
This is the core of this PR - only check for missing headers from before the validation window. This assumes that either the validator is currently backfilling headers in that window, or else the header in question will fall out of the validation window and be picked up by a subsequent iteration of the backfill process.

Reply:
Thanks for this synopsis! 🙏